src/AppInstallerRepositoryCore/Microsoft/SQLiteIndexSource.cpp

```diff
@@ -288,14 +288,34 @@ namespace AppInstaller::Repository::Microsoft
             std::vector<PackageVersionKey> GetAvailableVersionKeys(PinBehavior) const override
             {
                 std::shared_ptr<SQLiteIndexSource> source = GetReferenceSource();
-                std::vector<Utility::VersionAndChannel> versions = source->GetIndex().GetVersionKeysById(m_idId);
 
-                std::vector<PackageVersionKey> result;
-                for (const auto& vac : versions)
                 {
-                    result.emplace_back(source->GetIdentifier(), vac.GetVersion().ToString(), vac.GetChannel().ToString());
+                    auto sharedLock = m_availableVersionKeysLock.lock_shared();
+
+                    if (!m_availableVersionKeys.empty())
+                    {
+                        return m_availableVersionKeys;
+                    }
                 }
-                return result;
+
+                auto exclusiveLock = m_availableVersionKeysLock.lock_exclusive();
+
+                if (!m_availableVersionKeys.empty())
+                {
+                    return m_availableVersionKeys;
+                }
+
+                std::vector<SQLiteIndex::VersionKey> versions = source->GetIndex().GetVersionKeysById(m_idId);
+
+                for (const auto& vk : versions)
+                {
+                    std::string version = vk.VersionAndChannel.GetVersion().ToString();
+                    std::string channel = vk.VersionAndChannel.GetChannel().ToString();
+                    m_availableVersionKeys.emplace_back(source->GetIdentifier(), version, channel);
+                    m_availableVersionKeysMap.emplace(MapKey{ std::move(version), std::move(channel) }, vk.ManifestId);

```

Comment:
Why does this one not take the source ID?

Reply:
It isn't needed for the mapping.